### PR TITLE
GH-539 CONV_P - Adds publish params for already existing functions

### PR DESCRIPTION
**Pull request by mlzummo**
See https://github.com/hpcc-systems/Visualization/pull/596 for full list of comments.

### CODE REVIEW

**Comment by mlzummo** [comment]
fixing new found issues one moment ....

File: src/common/ResizeSurface.js [hunk 1]
```diff
@@ -12,7 +12,7 @@
         this.handleWidth = 8;
         this.handles = [{ loc: "NW" }, { loc: "N" }, { loc: "NE" }, { loc: "E" }, { loc: "SE" }, { loc: "S" }, { loc: "SW" }, { loc: "W" }];
 
-        this._allowResize = true;
+        this.allowResize(true);
```
Comment (GordonSmith) [inline src/common/ResizeSurface.js]: Not needed.

**Comment by GordonSmith** [comment]
Needs rebase.

**Comment by GordonSmith** [comment]
See https://github.com/hpcc-systems/Visualization/pull/596 for full list of comments.

**Comment by mlzummo** [comment]
@GordonSmith How are we looking with the new publish mods that will allow me to refactor this a bit better?

**Comment by GordonSmith** [comment]
I can probably put one of them into Util for you now.  Need to re-review it...

**Comment by mlzummo** [comment]
@GordonSmith ok sounds good just issue a PR against my fork/branch for util ... is util ok now?

**Comment by mlzummo** [comment]
@GordonSmith Status Update ?

**Comment by mlzummo** [comment]
@GordonSmith rebasing/reissueing (1.6.2) did not help ...

**Comment by mlzummo** [comment]
@GordonSmith Please Review and PULL

File: src/chart/Pie.js [hunk 2]
```diff
@@ -37,6 +33,9 @@
 
     Pie.prototype.publish("paletteID", "default", "set", "Palette ID", Pie.prototype._palette.switch(),{tags:["Basic","Shared"]});
     Pie.prototype.publish("useClonedPalette", false, "boolean", "Enable or disable using a cloned palette",null,{tags:["Intermediate","Shared"]});
+    Pie.prototype.publish("outerText", false, "boolean", "Sets label position inside or outside chart",null,{tags:["Basic"]});
+    Pie.prototype.publish("radius", 100, "number", "Sets radius in pixels of pie chart",null,{tags:["Basic"]});
+    Pie.prototype.publish("innerRadius", 0, "number", "Sets radius in pixels of center hole in pie chart (donut)",null,{tags:["Basic"]});
```
Comment (GordonSmith) [inline src/chart/Pie.js]: publishing radius and innerRadius is a bit misleading as they just get overridden when the Pie chart gets sized?  I think innerRadius should be %?

File: src/chart/Pie.js [hunk 3]
```diff
@@ -25,8 +21,8 @@
             .value(function (d) { return d[1]; })
         ;
         this.d3Arc = d3.svg.arc()
-            .padRadius(this._radius)
```
Comment (GordonSmith) [inline src/chart/Pie.js]: This is needed

File: src/common/Surface.js [hunk 4]
```diff
@@ -53,25 +51,36 @@
...
+    Surface.prototype.publish("showContent", true, "boolean", "Show Content",null,{tags:["Intermediate"]});
+    Surface.prototype.publish("menu", [], "array", "Menu List Data",null,{tags:["Intermediate"]});
+    Surface.prototype.publish("menuPadding", 2, "number", "Menu Padding",null,{tags:["Advanced"]});
+
+    Surface.prototype._menuParam = Surface.prototype.menu;
     Surface.prototype.menu = function (_) {
-        if (!arguments.length) return this._menu.data();
-        this._menu.data(_);
-        return this;
+        Surface.prototype._menuParam.apply(this, arguments);
```
Comment (GordonSmith) [inline src/common/Surface.js]: should be var retVal = ...

File: src/common/Surface.js [hunk 5]
```diff
@@ -53,25 +51,36 @@
...
     Surface.prototype.showContent = function (_) {
-        if (!arguments.length) return this._showContent;
-        this._showContent = _;
-        if (this.content()) {
-            this.content().visible(this._showContent);
+        var retVal = Surface.prototype._showContent.apply(this, arguments);
+        if (arguments.length) {
+            if (this.content()) {
+                this.content().visible(this.showContent());
+                return this;
```
Comment (GordonSmith) [inline src/common/Surface.js]: ```return this;``` not needed.

File: src/map/Choropleth.js [hunk 6]
```diff
@@ -26,6 +26,7 @@
 
     Choropleth.prototype.publish("paletteID", "YlOrRd", "set", "Palette ID", Choropleth.prototype._palette.switch(),{tags:["Basic","Shared"]});
     Choropleth.prototype.publish("useClonedPalette", false, "boolean", "Enable or disable using a cloned palette",null,{tags:["Intermediate","Shared"]});
+    Choropleth.prototype.publish("projection", "", "String", "Map project type",null,{tags:["Intermediate","Shared"]});
```
Comment (GordonSmith) [inline src/map/Choropleth.js]: Invalid type "String" should be "set"

File: src/other/Audio.js [hunk 7]
```diff
@@ -11,18 +11,13 @@
...
     Audio.prototype = Object.create(HTMLWidget.prototype);
     Audio.prototype.constructor = Audio;
     Audio.prototype._class += " other_Audio";
 
-    Audio.prototype.source = function (_) {
-        if (!arguments.length) return this._source;
-        this._source = _;
-        return this;
-    };
+    Audio.prototype.publish("source", [], "array", "Audio Source");
```
Comment (GordonSmith) [inline src/other/Audio.js]: Is not an array

File: src/other/MorphText.js [hunk 8]
```diff
@@ -1,22 +1,24 @@
...
-        this._reverse = false;
     }
+
     MorphText.prototype = Object.create(SVGWidget.prototype);
     MorphText.prototype.constructor = MorphText;
     MorphText.prototype._class += " other_MorphText";
 
+    MorphText.prototype.publish("anchor","middle", "set", "Sets anchor point",["middle"],{tags:["Basic"]});
+    MorphText.prototype.publish("fontSize",14, "number", "Sets fontsize",null,{tags:["Basic"]});
+    MorphText.prototype.publish("reverse",false, "boolean", "Reverse Text",null,{tags:["Basic"]});
```
Comment (GordonSmith) [inline src/other/MorphText.js]: "Reverse Animation"

**Comment by GordonSmith** [comment]
can you rename all the proxied functions to:
```
  _origXXX
```
to make them consistent with the other instances in the framework.

**Comment by mlzummo** [comment]
@GordonSmith do you know of a regex or a way i can search for all proxied functions without reading every function of every single files?

**Comment by mlzummo** [comment]
besides this which im about to use: .prototype._

**Comment by GordonSmith** [comment]
I am only referring to your proxies in this change which are either using:
._XXX
or 
.XXX_orig

I have been trying to standardize on ._origXXX

File: src/other/Audio.js (hunk 7 above)
Comment (GordonSmith) [inline src/other/Audio.js]: A string (well its an URL really)

File: src/chart/Pie.js (hunk 2 above)
Comment (GordonSmith) [inline src/chart/Pie.js]: If I set the radius to be 200 and I resize the page, what radius do expect the Pie to have afterwards?

File: src/other/MorphText.js [hunk 9]
```diff
@@ -1,22 +1,24 @@
...
 (function (root, factory) {
     if (typeof define === "function" && define.amd) {
-        define(["../common/SVGWidget", "css!./MorphText"], factory);
+        define(["../common/SVGWidget", "../api/I1DChart", "css!./MorphText"], factory);
     } else {
-        root.other_MorphText = factory(root.common_SVGWidget);
+        root.other_MorphText = factory(root.common_SVGWidget, root.api_I1DChart);
     }
-}(this, function (SVGWidget) {
+}(this, function (SVGWidget, I1DChart) {
```
Comment (GordonSmith) [inline src/other/MorphText.js]: Remove I1DChart import.

File: src/chart/Pie.js [hunk 10]
```diff
@@ -84,6 +58,8 @@
             this._palette = this._palette.cloneNotExists(this.paletteID() + "_" + this.id());
         }
 
+        this.d3Arc.innerRadius(this.width() * this.innerRadius() ? this.innerRadius() : 0);
```
Comment (GordonSmith) [inline src/chart/Pie.js]: What is this line doing?

Comment (mlzummo) [inline src/chart/Pie.js]: @GordonSmith that is where the innerradius (pie whole is adjusted) ... if its not updated there it isn't updated correctly ... we had discussed that it needs to go in the update function ... the radius param is now a % as in % of the width of the circle (i dont think i need to divide by 2 do i?)

File: src/chart/Pie.js (hunk 3 above)
Comment (mlzummo) [inline src/chart/Pie.js]: its not showing it was updated but the code has been updated

File: src/chart/Pie.js (hunk 10 above)
Comment (GordonSmith) [inline src/chart/Pie.js]: In which case it should be more like this (off the top of my head):
```
this.d3Arc.innerRadius(this.innerRadius_exists() ? (this.width() / 2) * (this.innerRadius() / 100) : 0);
```

File: src/chart/Pie.js [hunk 11]
```diff
@@ -37,33 +33,11 @@
 
     Pie.prototype.publish("paletteID", "default", "set", "Palette ID", Pie.prototype._palette.switch(),{tags:["Basic","Shared"]});
     Pie.prototype.publish("useClonedPalette", false, "boolean", "Enable or disable using a cloned palette",null,{tags:["Intermediate","Shared"]});
+    Pie.prototype.publish("outerText", false, "boolean", "Sets label position inside or outside chart",null,{tags:["Basic"]});
+    Pie.prototype.publish("innerRadius", 0, "number", "Sets radius in pixels of center hole in pie chart (donut)",null,{tags:["Basic"]});
```
Comment (GordonSmith) [inline src/chart/Pie.js]: Description is wrong

File: src/chart/Pie.js [hunk 12]
```diff
@@ -84,6 +58,8 @@
             this._palette = this._palette.cloneNotExists(this.paletteID() + "_" + this.id());
         }
 
+        this.d3Arc.innerRadius(this.innerRadius_exists() ? (this.calcRadius()) * (this.innerRadius() / 100) : 0);
```
Comment (GordonSmith) [inline src/chart/Pie.js]: Change to:
```
this.calcRadius() * this.innerRadius() / 100
```

**Pull request merged**